oprf-service/src/services/crypto_device/zk.rs

```diff
@@ -0,0 +1,193 @@
+use std::ops::Shr;
+use std::{collections::HashMap, path::Path, sync::Arc};
+
+use ark_bn254::Bn254;
+use ark_ff::{AdditiveGroup, BigInt, Field as _, UniformRand as _};
+use circom_types::{groth16::ZKey, traits::CheckElement};
+use eyre::Context as _;
+use groth16::{CircomReduction, ConstraintMatrices, Groth16, Proof, ProvingKey};
+use itertools::{Itertools as _, izip};
+use oprf_core::keys::keygen::KeyGenPoly;
+use oprf_types::crypto::{RpSecretGenCiphertext, RpSecretGenCiphertexts};
+use tracing::instrument;
+use witness::{BlackBoxFunction, Graph, ruint::aliases::U256};
+
+use crate::services::crypto_device::CryptoDevice;
+
+// TODO maybe we can unify this with the OPRF-client material?
+pub(super) struct Groth16Material {
+    pk: ProvingKey<Bn254>,
+    matrices: ConstraintMatrices<ark_bn254::Fr>,
+    graph: Graph,
+}
+
+impl Groth16Material {
+    pub(crate) fn from_paths(
```

Comment:
Would suggest having a from reader variant that can take bytes directly instead or at least in addition.